Fetch meta fields in FetchFieldsPhase using ValueFetcher

salvatore-campagna opened the pull request:
Here we extract the logic to populate metadata fields such as `_ignored`, `_routing`, `_size` and the deprecated `_type` into `FetchFieldsPhase` so that we can use the `ValueFetcher` interface to retrieve field values. This allows us to fetch values no matter if the Mapper uses stored or doc values.

This is a refactoring required for #101373

salvatore-campagna committed:
refactor: extract meta fields fetching in separate fetch sub-phase

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

elasticsearchmachine commented:
Pinging @elastic/es-storage-engine (Team:StorageEngine)

salvatore-campagna committed:
fix: inline return value of storedFieldSpec

salvatore-campagna committed:
fix: check stored fields not fetch fields

javanna commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/StoredFieldsPhase.java:
```diff
@@ -77,7 +68,7 @@ public FetchSubPhaseProcessor getProcessor(FetchContext fetchContext) {
                     for (String fieldName : fieldNames) {
                         MappedFieldType ft = sec.getFieldType(fieldName);
-                        if (ft.isStored() == false) {
+                        if (ft.isStored() == false || sec.isMetadataField(fieldName)) {
```
Given that StoredField is a private record, and we create it within this class, I think we should simply skip creating an instance for metadata fields, instead of creating it and later ignoring it?

salvatore-campagna replied:
If the field is a metadata field we hit the if and execute `continue` which means we do not create the stored field and we do not add it to `storedFields`. So I think we are not creating an instance od `StoredField` for metadata fields. So it looks to me that we already skip the creation of the stored field. Am I wrong?

I guess one could argue that later the check on `storedField.isMetadataField == false` is not necessary because of this since all fields we are are not metadata fields, right?

javanna replied:
You are correct, I think we should now be able to remove the isMetadataField member from the private StoredField record. There is also a conditional to exclude the `_source` from retrieval, that can and should be now removed.

javanna commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchMetadataFieldsPhase.java:
```diff
@@ -0,0 +1,72 @@
+    @Override
+    public FetchSubPhaseProcessor getProcessor(FetchContext fetchContext) {
+        StoredFieldsContext storedFieldsContext = fetchContext.storedFieldsContext();
+        if (storedFieldsContext == null || storedFieldsContext.fetchFields() == false) {
```
I see that MetadataFetcher below uses valueFetchers to retrieve values, and that is great. That means that this phase should no longer be exposed to how a certain field is stored, it just retrieves metadata fields via value fetchers. 

That should be straightforward for metadata fields which get returned by default, because they are not explicitly requested and instead of being retrieve from stored fields they go via value fetchers.

Metadata fields that are requested explicitly are requested in different ways: via `fields`, which goes through `FetchFieldsPhase`, and via `stored_fields` which currently goes through `StoredFieldsPhase`. It seems to me that you intend to move the latter and not the former to this new phase. In fact `FetchFieldsPhase` already goes through value fetchers, so requesting `fields:_ignored` would become equivalent to retrieving it via `stored_fields:_ignored` in how the field is retrieved, but it would be returned in a different section in the search response?

What is odd here is that for this phase to return metadata fields via value fetchers, they need to be requested via `stored_fields`. That tells me that it may make sense to unify this new phase with the existing FetchFieldsPhase, and add the notion of extracting meta fields requested via stored fields context. That would also require less code duplication, or am I missing something?

javanna commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchMetadataFieldsPhase.java:
```diff
@@ -0,0 +1,72 @@
+        };
+    }
+
+    private static List<FieldAndFormat> getAdditionalFields(final StoredFieldsContext storedFieldsContext) {
```
I don't understand the purpose of this method, shouldn't we extract only metadata fields that are explicitly requested via `stored_fields`. What do we do instead? Why the explicit mention of `_size`?

javanna commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/MetadataFetcher.java:
```diff
@@ -0,0 +1,112 @@
+import java.util.Map;
+import java.util.stream.Stream;
+
+public class MetadataFetcher {
```
given that this is the only way to retrieve metadata fields in the new phase, I don't see why we need a separate component for it. A big portion of it is what FetchFieldsPhase already does, so maybe we should unify the two and try to reuse what is already available? What is the need for a new component / phase ? The only difference is where fields are returned in the search response?

javanna reviewed (commented):
Thanks for opening this separate PR, I left some comments.

salvatore-campagna committed:
refactor: just use FetchFieldsPhase to fetch both fields and stored_fields

salvatore-campagna committed:
fix: update skip version

salvatore-campagna committed:
fix: simplify code

salvatore-campagna committed:
fix: prevent a NPE when retrieving the field type for metadata fields

salvatore-campagna commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java:
```diff
@@ -45,6 +45,7 @@ public static FieldFetcher create(SearchExecutionContext context, Collection<Fie
         List<String> unmappedFetchPattern = new ArrayList<>();
         List<ResolvedField> resolvedFields = new ArrayList<>();
 
+        context.setAllowedFields(context::isMetadataField);
```
I had to add this here because otherwise some tests fail because a NPE. The NPE happens later on when in FIeldFetcher we do

```
MappedFieldType ft = context.getFieldType(field);
```

this would return null for metadata fields and will result in an NPE when calling


```
return fieldAndFormat.ft.valueFetcher(context, fieldAndFormat.format);
```
in buildValueFetcher.

salvatore-campagna replied:
The explanation I have for this is that for metadata fields which are returned by default we need them to be "allowed". Honestly it is not really clear to me what "being allowed" means in this context...I can only see the flag `allowedFieields` (predicate) is used in the `QueryRewriteContext#fieldType` method which I interpret that "getting the type for a field might be allowed or not"...but I don't know the reason why.

salvatore-campagna replied:
I reverted the change because it was breaking more tests then it was fixing.

salvatore-campagna committed:
Revert "fix: prevent a NPE when retrieving the field type for metadata fields"

This reverts commit 99f129d1142a47762a18a2131c4bd12e2e602da3.

salvatore-campagna committed:
fix: fetch _source using ValueFetcher

salvatore-campagna committed:
fix: allow _ignored, _routing and _type fields

If fields are not allowed, later on we are not able to retrieve the
corresponding MappedFieldType which results in a null pointer exception.
Filtering such fields happens by means of the QueryRewriteContext#allowedFields
which prevents fetching mappings if the field predicate evaluates
to false.

salvatore-campagna committed:
fix: return stored fields with their correct type not as strings

salvatore-campagna committed:
fix: process just metadata fields

salvatore-campagna committed:
fix: restore original exception message

salvatore-campagna committed:
fix: hadling of alias fields

salvatore-campagna committed:
nit: improve readability

salvatore-campagna committed:
fix: mock MappedFieldType#name

salvatore-campagna committed:
fix: add a missing FetchFieldsPhase

salvatore-campagna committed:
fix: use a visitor per each field instead of per each doc

salvatore-campagna committed:
test: include a test to surface a bug

The issue happens when we try to use multiple stored fields since
we have a single shared instance of SingleFieldsVisitor per document.

salvatore-campagna committed:
Merge branch 'main' into refactor/meta-fields-sub-fetch

salvatore-campagna committed:
fix: merge conflict mistake

salvatore-campagna committed:
fix: remove empty line

salvatore-campagna committed:
fix: remove empty line

salvatore-campagna committed:
fix: include a boolean to enable fetching of stored fields in FieldFetcher

salvatore-campagna committed:
fix: handle _size metadata field

salvatore-campagna committed:
fix: move the fetch fase earlier

salvatore-campagna committed:
fix: call setNextReader before setting field loaders

salvatore-campagna committed:
fix: skip unmapped fields

salvatore-campagna committed:
fix: load directly stored fields if they are not pre-loaded

salvatore-campagna committed:
fix: add missing return

javanna commented on server/src/main/java/org/elasticsearch/search/SearchModule.java:
```diff
@@ -1014,10 +1014,10 @@ private <T extends QueryVectorBuilder> void registerQueryVectorBuilder(QueryVect
         registerFetchSubPhase(new FetchDocValuesPhase());
         registerFetchSubPhase(new ScriptFieldsPhase());
         registerFetchSubPhase(new FetchSourcePhase());
-        registerFetchSubPhase(new FetchFieldsPhase());
```
is this related to the NPE in #106617 ? I'd like to better understand where we rely on ordering and why that needs changing.

salvatore-campagna replied:
I reverted this...

javanna commented on server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java:
```diff
@@ -128,15 +128,15 @@ private SearchHits buildSearchHits(SearchContext context, int[] docIdsToLoad, Pr
                 fieldLookupProvider.setNextReader(ctx);
-                for (FetchSubPhaseProcessor processor : processors) {
-                    processor.setNextReader(ctx);
-                }
```
What issue made you move this above?

salvatore-campagna replied:
I reverted this too

javanna commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java:
```diff
@@ -54,16 +71,27 @@ public static FieldFetcher create(SearchExecutionContext context, Collection<Fie
 
             for (String field : context.getMatchingFieldNames(fieldPattern)) {
                 MappedFieldType ft = context.getFieldType(field);
+                if (ft == null) { // unmapped field
```
why is this new conditional required only now?

salvatore-campagna replied:
I saw failures for this already before in some local tests and I am not sure why I didn't see them in the CI...I guess again some tests are not executed if there are other tests failing. This happens for some fields

salvatore-campagna replied:
That chack is required because later on we do `ft.isStored()` which we need to do because we need to check if we need to fetch fields that are part of `stored_fields`

salvatore-campagna replied:
Notice that the behavior of `FecthFields` or put it another way, the way of handling `fields` is described by the test SizeMappingIT like

```
assertResponse(
            prepareSearch("test").addFetchField("*"),
            response -> assertNull(response.getHits().getHits()[0].getFields().get("_size"))
        );
```

which is...if a wildcard is used we **do not return** metadata fields such as `_size`.

salvatore-campagna replied:
Come se non bastassea quando i field sono richiesti attraverso `stored_fields` la chiamata a `context.getFieldType(field)` restituisce null come `MappedFieldType`. Questo pero' non e' un problema perche' in quel caso quei field saranno recuperati attraverso la `StoredFieldPhase` e quindi il `ValueFecther` non verra' usato...in effetti non possiamo usarlo in quel caso perche' `MappedFiledType` e' **null**.

salvatore-campagna replied:
In summary:
1. We need to check for `MappedFieldType` to be `null` because when handling fields passed as `stored_fields` in the `StoredFieldContext` we need to skip **some of them**...and we can't do otherwise because in that case `MappedFieldType` is null.
2. We can't add a ft = null to the list of resolved fields later on in the loop because otherwise later on sorting of resolved fields breaks with a NPE. See `resolvedFields.sort(Comparator.comparing(resolvedField -> resolvedField.mappedFieldType.name()));`

salvatore-campagna replied:
I honestly don't know why `context.getFieldType(field)` is null when handling stored fields...maybe there is no MapperService available?

javanna commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/UserBoolQueryBuilder.java:
```diff
@@ -30,6 +33,13 @@ public class UserBoolQueryBuilder extends BoolQueryBuilder {
+        IgnoredFieldMapper.NAME,
+        LegacyTypeFieldMapper.NAME,
+        "type"
+    );
```
Can you describe the problem you encountered that made you add two new lists with the same metadata field names? I guess that the stored fields phase was previously skipping these checks that the fetch fields phase now bumps into?

salvatore-campagna replied:
I might actually need to remve tese from the list after I introduced `fetchStoedFields` in `FieldFetcher`

salvatore-campagna replied:
Anyway the reason behind it is this, I will double check if that is strictly required since after introducing `fetchStoredFields` and checking `ft == null` in `FieldFetcher` this might not be needed anymore (as a side effect).

```
// If the field is not allowed, behave as if it is not mapped
        if (allowedFields != null && false == allowedFields.test(name)) {
            return null;
        }
```

Some security related queries like UserBookQuery and ApiBoolQuery use this setting to allow/disallow fields.
The `MappedFieldType` is not available in that case which means a `ValueFecther` is not available.

salvatore-campagna replied:
Honestly I find this way of handling allowed fields quite obscure and definitely not easy to get until you step on a bug or test failure.

salvatore-campagna replied:
This was added in this pr [8c09fc8f9fa9e9c75521491594901fc3b3e6023c](https://github.com/elastic/elasticsearch/commit/8c09fc8f9fa9e9c75521491594901fc3b3e6023c)

salvatore-campagna replied:
Maybe it would have been better to return a subclass of `MappedFieldType`...something like `SecurityHiddenMappedFieldType` which might return null/empty values from its `ValueFetcher` to hide details that are expected to be hidden?

salvatore-campagna replied:
In summary this might not be required anymore because when `allowedFields` is used the `MappedFieldType` is null as explained above and we skip those fields in the `FieldFetcher`.

salvatore-campagna commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java:
```diff
@@ -47,12 +71,20 @@ public StoredFieldsSpec storedFieldsSpec() {
+     * @return if we have to include the _size field or not
+     */
+    private static boolean includeSizeMetadataField(final List<FieldAndFormat> fields) {
+        return fields.stream().map(field -> field.field).noneMatch("*"::equals);
```
I "extracted" this behavior from test `SizeMappingIT#testGetWithFields` which explicitly tests this behaviour testing `_size` requested explicitly or with a wildcard `*` for `fields` and `stored_fields`.

javanna replied:
I mentioned it above, I think this is problematic and we should have no mention of `_size` in here, as it is a custom metadata field that a plugin can provide. We need to be able to detect custom metadata fields in a generic manner, without knowing their names, otherwise they are not pluggable.

salvatore-campagna committed:
fix: revert FetchPhase ordering and setNextReader

salvatore-campagna committed:
Merge branch 'main' into refactor/meta-fields-sub-fetch

salvatore-campagna committed:
fix: revert changes to get stack traces

salvatore-campagna committed:
fix: clarify comment

When fetching fields through `stored_fields` the `MappedFieldType`
is not available so we can't use the `ValueFetcher`. This is not an
issue since those fields are handled by the `StoredFieldsPhase`.

salvatore-campagna committed:
fix: clarify null check for unmapped fields

salvatore-campagna committed:
fix: use the actual field name for sorting

salvatore-campagna committed:
Revert "fix: use the actual field name for sorting"

This reverts commit 7dcf8cef9ea3feff5dce490bfbda67655af0b7c7.

salvatore-campagna committed:
Merge branch 'main' into refactor/meta-fields-sub-fetch

salvatore-campagna committed:
fix: possible NPE with storedFields

salvatore-campagna committed:
fix: clarify why we skip null MappedFieldType

salvatore-campagna committed:
checkstyle: remove implmnote

salvatore-campagna committed:
fix: do not try fetching stored fields if there is none

salvatore-campagna committed:
fix: revert storedFields == null check

salvatore-campagna committed:
Revert "fix: revert storedFields == null check"

This reverts commit 60c4d9ee37645aa1de96500f0928ea501a88907a.

salvatore-campagna committed:
Revert "fix: do not try fetching stored fields if there is none"

This reverts commit 7092e1c3bbd75a6f3608fd2c63e0d57f46db7bdb.

salvatore-campagna committed:
fix: require non-null loadedFields

salvatore-campagna committed:
nit: remove comment

salvatore-campagna committed:
fix: remove storedFields null check and disable parallel execution for ChildrenAggregationBuilder

salvatore-campagna committed:
Merge branch 'main' into refactor/meta-fields-sub-fetch

salvatore-campagna committed:
fix: disallow top hits parallel execution

TopHits executes the fetch phase as part of the query phase when calling
builAggregation. As a result the fetch phase is executed concurrently
by, possibly, multiple search_worker threads. The phase phase code
is not thred safe whn calling setNextReader and nextDoc so we disable
parallel execution of TopHitsAggregator.

salvatore-campagna committed:
fix: null check not required after #106862

salvatore-campagna committed:
Merge branch 'main' into refactor/meta-fields-sub-fetch

javanna commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/StoredFieldsPhase.java:
```diff
@@ -98,18 +89,12 @@ public void setNextReader(LeafReaderContext readerContext) {
                     }
                 }
-                hitContext.hit().addDocumentFields(docFields, metaFields);
+                hitContext.hit().addDocumentFields(docFields, Collections.emptyMap());
```
use setDocumentField instead? we don't need the map at all at this point, you can simply call the method from within the loop for each field.

javanna commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java:
```diff
@@ -9,30 +9,54 @@
+
+    public static boolean isMetadataField(final String field) {
+        return METADATA_FIELDS.stream().map(fieldAndFormat -> fieldAndFormat.field).anyMatch(fieldName -> fieldName.equals(field));
+    }
```
I think that this is problematic, because you have a separate definition of metadata fields within fetch fields. Plugins (like `SizeMapperPlugin`) can provide their own extensions of custom metadata fields. When they do so, we should treat those metadata fields like any other. We certainly should not have the knowledge of which metadata fields plugins may provide. Ideally, we would simply rely on `SearchExecutionContext#isMetadataField` here, like `StoredFieldsPhase` used to do. Additionally , you need to skip the `_source` field exactly like `StoredFieldsPhase` used to do.

javanna commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java:
```diff
@@ -47,12 +71,20 @@ public StoredFieldsSpec storedFieldsSpec() {
-                for (Map.Entry<String, DocumentField> entry : documentFields.entrySet()) {
-                    hit.setDocumentField(entry.getKey(), entry.getValue());
-                }
+                final FieldFetcher.DocAndMetaFields fields = fieldFetcher.fetch(hitContext.source(), hitContext.docId());
```
In order to simplify things, I would propose that we make no changes to `FieldFetcher`, and instead create two instances of it, one for ordinary fields (like we already do) and a second one for metadata fields. The choice of where to put the fields that each fetcher outputs can be made in `FetchFieldsPhase`. This simplifies things in that it removes the need to distinguish between metadata and not in FieldFetcher and removes the need for the `fetchStoredFields` flag too.

javanna reviewed (commented):
left some more comments

salvatore-campagna committed:
fix: remove isMetadataField from StoredField

salvatore-campagna committed:
fix: remove redundant check on field name

salvatore-campagna committed:
fix: use setDocumentField

salvatore-campagna committed:
fix: remoe logic behind fetching _size

salvatore-campagna committed:
fix: reuse FieldFetcher as it was originally

salvatore-campagna committed:
fix: only deal with non-metadata fields that are stored

salvatore-campagna committed:
fix: remove unused method

salvatore-campagna committed:
fix: flip merge call

salvatore-campagna committed:
fix: just skip fields not stored

salvatore-campagna commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/370_profile.yml:
```diff
@@ -163,7 +169,8 @@ disabling stored fields removes fetch sub phases:
   - match: { profile.shards.0.fetch.debug.stored_fields: [] }
-  - is_false: profile.shards.0.fetch.children
+  - length: { profile.shards.0.fetch.children: 1 }
+  - match: { profile.shards.0.fetch.children.0.type: FetchFieldsPhase }
```
That was the purpose of the `fetchStoredFields` flag.

salvatore-campagna replied:
That would break another test in `370_profile` when checking

```
- match: { profile.shards.0.fetch.debug.stored_fields: [_id, _routing, _source] }
```

javanna commented:
Additional thought: can you check that we have a test that verifies that _ignored as well as other metadata fields are returned when using the get API? We are focusing on the search API for now, and its fetch phase, but stored fields are retrieved directly via `ShardGetService` too.

salvatore-campagna committed:
fix: hanlde none stored_fields

salvatore-campagna committed:
fix: do not include metadata stored fields specs

salvatore-campagna commented on modules/parent-join/src/main/java/org/elasticsearch/join/aggregations/ChildrenAggregationBuilder.java:
```diff
@@ -123,6 +124,11 @@ protected XContentBuilder doXContentBody(XContentBuilder builder, Params params)
     }
 
+    @Override
+    public boolean supportsParallelCollection(ToLongFunction<String> fieldCardinalityResolver) {
```
I will remove this after merging the other pr.

salvatore-campagna committed:
fix: fetch fields phase not executed if no fields or stored fields

salvatore-campagna committed:
Merge branch 'main' into refactor/meta-fields-sub-fetch

salvatore-campagna committed:
revert: parallel execution of top hits and child agg

salvatore-campagna committed:
revert: Objects.requireNotNull

salvatore-campagna committed:
test: yaml test _ignored stored_fields in get api

salvatore-campagna committed:
fix: yaml test header

javanna reviewed (commented):
I left some comments

salvatore-campagna commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/get/120_stored_fields_ignored.yml:
```diff
@@ -0,0 +1,166 @@
+ip:
+  - skip:
+      version: " - 8.4.99"
+      reason: ip introduced in 8.5.0
```
This is when support for `ignore_mallformed` was added for the ip field type...I use an ip address later on other than a keyword and a numeric type.

salvatore-campagna commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java:
```diff
@@ -9,35 +9,55 @@
+            : fetchFieldsContext.fields() == null ? Collections.emptyList()
+            : fetchFieldsContext.fields();
+
+        final FieldFetcher fieldFetcher = FieldFetcher.create(fetchContext.getSearchExecutionContext(), fetchFields);
```
Yeah we can avoid that...but then I also have to start doing all kind of null checks...

javanna replied:
yes, null checks are the downside, but at least we avoid doing needless work and creating needless instances of fetchers.

salvatore-campagna committed:
fix: remove unnecessary skip

salvatore-campagna committed:
fix: avoid creating a field fetcher if possible

salvatore-campagna committed:
fix: remove unecessary method mock

salvatore-campagna commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java:
```diff
@@ -9,35 +9,55 @@
  */
 public final class FetchFieldsPhase implements FetchSubPhase {
+
+    private static final List<FieldAndFormat> METADATA_FIELDS = List.of(
```
id `DEFAULT_METADATA_FIELDS` ok?

javanna replied:
sure

salvatore-campagna commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java:
```diff
@@ -9,35 +9,55 @@
+            : fetchFieldsContext.fields();
+
+        final FieldFetcher fieldFetcher = FieldFetcher.create(fetchContext.getSearchExecutionContext(), fetchFields);
+        final FieldFetcher metadataFieldFetcher = FieldFetcher.create(fetchContext.getSearchExecutionContext(), METADATA_FIELDS);
```
Sorry I read this multiple times but I am not sure I get what you means:
* which test is missing? one where both `fields` and `stored_fields` are none?
* I don't get your second comment can you clarify?

javanna replied:
> which test is missing? one where both fields and stored_fields are none? My understanding is that only stored_fields accepts "none" as an option.

No worries, let me clarify: one that requests some fields with `fields` and `stored_fields: _none_` . I believe that when fields are requested the fields context will be not null, hence the phase will execute, in which case you need to check again if fetching stored fields was disabled, or you end up returning the default metadata fields that should not be returned? We should test that and it surprises me that there are no failures. Does this make sense?

My second comment revolves around the fact that you use the stored fields context only to disable loading of default metadata fields, but other metadata fields requested via `stored_fields` are still retrieved via `StoredFieldsPhase`. Also, if you request `stored_fields: _ignored`, I believe that it's again retrieved from stored fields directly without going through value fetchers, which will break once you move the field to have doc_values. In a previous iteration you were adding metadata fields from stored fields context and retrieving them in FetchFieldsPhase. You also had a conditional in StoredFieldsPhase that excluded metadata fields from it. Without this latter conditional, you are now removing the isMetadataField flag from StoredField, but don't you still handle metadata fields in `StoredFieldsPhase` when they are explicitly requested? Those should go via value fetchers as well. I think that with the flag removed the may end up among document fields in the response, but I am not entirely sure, and weird that there are no test failures around it. Does this make sense?

salvatore-campagna replied:
My understanding is that we can't fetch some metadata fields via FetchFieldsPhase because some metadata fields are missing a `ValueFetcher` implementation. Doing that would result in throwing an `UnsupportedOperationException`. Am I wrong?

`SeqNoFieldMapper` does this

```
public ValueFetcher valueFetcher(SearchExecutionContext context, String format) {
     throw new UnsupportedOperationException("Cannot fetch values for internal field [" + name() + "].");
}
```

(note also that it return a 5xx error...which is also an issue in serverless because that is mapped to a "Service Unavailble" error).

salvatore-campagna commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/StoredFieldsPhase.java:
```diff
@@ -68,21 +56,19 @@ public FetchSubPhaseProcessor getProcessor(FetchContext fetchContext) {
-                        storedFields.add(new StoredField(fieldName, ft, sec.isMetadataField(ft.name())));
-                        fieldsToLoad.add(ft.name());
+                Collection<String> fieldNames = sec.getMatchingFieldNames(field);
+                for (String fieldName : fieldNames) {
```
The way it is written now means that this phase `StoredFieldPhase` is responsible for fetching stored fields (fields that have stored: true in the mapper) which are requested via `stored_fields` including when they are requested explicitly like in `stored_fields: [_one, _two]` or using `stored_fields: [ * ]`.

Stored fields which are requested via `fields` **must** be handled by FetchFieldsPhase.

salvatore-campagna replied:
This is the way the split should be done according to my udnerstanding:
1. stored fields requested via fields => FetchFieldsPhase using ValueFetcher
2. stored fields requested via stored_fields => **StoredFieldsPhase**
3. non-stored fields requested via fields +> FetchFieldsPhase using ValueFecther
4. non-stored fields requested via stored_fields => FetchFieldsPhase using ValueFetcher
5. stored metadata fields requested via fields => FetchFieldsPhase using ValueFetcher
6. stored metadata fields requested via stored_fields => **StoredFieldsPhase**
7. non-stored metadata fields requested via fields => FetchFieldsPhase using ValueFetcher
8. non-stored metadata fields requested via stored_fields FetchFieldsPhase using ValueFetcher

As you can see this is why I have written it that way and the way that makes `SizeMappingIT` work.

salvatore-campagna replied:
Also notice that this is the way that allows us to use the ValueFecther for **ALMOST** averything.

salvatore-campagna replied:
Also, as a result of removing `isMetadataField` the distinction on which phase to use does not depend anymore on whether the field is a metadata field or not but just on:
* field is stored or not in the mapper (`stored: true` or `stored: false`)
* field is requested via `fields` or `stored_fields`.

javanna replied:
> I think we need to agree on that because at the end the way I see it is that, at least from my perspective, it is not that important which phase is responsible for what as long as:

It is important once you move `_ignored` to have doc_values instead of being stored, it needs to go via value fetchers at all times, it can't be retrieved some times via `FetchFieldsPhase` and some other times via `StoredFieldsPhase`. I explained that in another comment above.


> Anyway I think the original idea behind the refactoring is: "we want to use ValueFetcher whenever possible" which means basically for all fields that have a ValueFetcher considering that some fields do not have it (they throw UnsupportedOperationException like for instance _seq_no).

That requires special treatment, indeed. The current distinction can be made using `MappedFieldType#isStored` , but once you have `_ignored` no longer stored, we need a different mechanism. We do need to exclude fields that have never been retrieved before. The reason is now that they are not stored, but that condition will need to be adapted. I tried different approaches around this: 
1) give those fields (e.g. _seq_no) a value fetcher that returns their data: we would make them retrievable via runtime fields as well which I am not sure is what we want.
2) give those fields an empty value fetcher, there are metadata fields that are currently excluded that already have a non empty value fetcher, hence they can be retrieved via script but are excluded from the `StoredFieldsPhase`
3) add a specific method to MappedFieldType: #107086 . I think this is probably the best approach. We can keep for now the ft.isStored check, and replace it later with a different conditional. This is not needed to get the current PR in.

javanna replied:
>  I think there should be tests that fail because of this, or maybe tests are missing.

let me clarify this too: I was under the impression that if `StoredFieldsPhase` still retrieves metadata fields, we must keep the isMetadataField flag in the StoredField record, or those metadata fields will be added among document fields and not meta fields in the response. I do see that SearchHit has a strange way of retrieving fields via get, that pulls from both doc as well as metadata fields (as a fallback). Maybe that covers issues. IN the output, I think that doc fields are serialized under `fields` in the response, while metadata fields at the hit, at the same level as `_source`. This distinction is not so clear when checking SearchHits via IT tests.

salvatore-campagna committed:
fix: fetch fields and stored fields

salvatore-campagna committed:
fix: always return default metadata fields

salvatore-campagna committed:
fix: remove unnecessary null check

salvatore-campagna committed:
fix: remove skip version

salvatore-campagna committed:
fix: remove reason as it is sligthly different for ccs

salvatore-campagna committed:
Merge branch 'main' into refactor/meta-fields-sub-fetch

salvatore-campagna committed:
fix: _ignored fetched through stored_fields

salvatore-campagna committed:
fix: do not fetch metadata fields in stored fields phase

javanna commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/get/120_stored_fields_ignored.yml:
```diff
@@ -0,0 +1,162 @@
+          - ops
+
+  - is_false: fields
+  - is_false: stored_fields
```
I wonder about this specific assertion: is `stored_fields` a section of the get API response?

javanna commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/520_fetch_fields.yml:
```diff
@@ -0,0 +1,142 @@
+  - match: { hits.hits.0.fields.value: null }
+
+---
+fetch _ignored via stored_fields:
```
That should not be the case if we make fetch fields phase fetch metadata fields that are requested via stored_fields (not only the default ones). I explained this [here](https://github.com/elastic/elasticsearch/pull/106325#discussion_r1553136287). I see that you addressed this in the code, so we should be ok, but we will have to address the `isStored` check in FetchFieldsPhase, for instance by merging #107086 or some similar solution.

javanna commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/520_fetch_fields.yml:
```diff
@@ -0,0 +1,142 @@
+  # This should be `unauthorized` (401) or `forbidden` (403) or at least `bad request` (400)
+  # while instead it is mapped to an `internal_server_error (500)`
+  - match: { status: 500 }
+  - match: { error.root_cause.0.type: unsupported_operation_exception }
```
Thanks for adding all these tests, they help a lot.

Can you add one as requested [here](https://github.com/elastic/elasticsearch/pull/106325#discussion_r1553136287), that requests `fields` as well as `stored_fields: _none_`? Could you check the current behaviour on it and make sure that stays the same after your change?

javanna replied:
the suggested additional test is redundant, like we've seen in the other comment below. We are good without.

javanna commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java:
```diff
@@ -10,48 +10,102 @@
+        if (storedFieldsContext != null
+            && storedFieldsContext.fieldNames() != null
+            && storedFieldsContext.fieldNames().isEmpty() == false) {
+            final Set<FieldAndFormat> metadataFields = new TreeSet<>(Comparator.comparing(fieldAndFormat -> fieldAndFormat.field));
```
why not use a list directly given that order matters? You wanted to remove duplicates?

salvatore-campagna replied:
When calling `getMatchingFieldNames` with a `*` it would return also fields like `_ignored`, `_routing` which would be added by default. At the end duplicates do not really cause an issue because a map used to add them to the document and the map eliminates duplicates because uses the name of the field as key. Note that this happens also in your branch...I verified it by actually running the code.

javanna replied:
sounds good, but with your comparator you change the order to be based on field name instead of insertion order? Maybe it is a good idea to remove duplicates, because otherwise we do end up fetching the same fields multiple times, although they get  de-duplicated later.

salvatore-campagna replied:
Ok I can replace the `TreeSet` with the `HashSet`.

javanna commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java:
```diff
@@ -10,48 +10,102 @@
+        final FieldFetcher metadataFieldFetcher;
+        if (storedFieldsContext != null
+            && storedFieldsContext.fieldNames() != null
+            && storedFieldsContext.fieldNames().isEmpty() == false) {
```
You still should check `storedFieldsContext.fetchFields()` here, or you are doing to end up fetching default metadata fields when `stored_fields: _none_` is requested (and fields are requested at the same time), which I think breaks backwards compatibility. Also, it would be inconsistent behaviour as `stored_fields: _none_` behaves differently depending on whether `fields` are requested or not.

Once you do that, you will need to add null checks around metadata field fetcher down below, like you do for the other fetcher already.

salvatore-campagna replied:
Ok so I understand that here there is a missing test which is one where `stored_fields: _none_` and `fields` requests a few fields.

salvatore-campagna replied:
I tested this on `main` before testing it on my branch

```
---
fetch fields with none stored_fields:
  - do:
      search:
        index: test
        body:
          stored_fields: _none_
          fields: [stored_keyword, keyword, stored_value, value, ignored_keyword, ignored_value, _ignored]

  - match: { hits.total.value: 1 }
```

Adding this to `520_fetch_fields.yaml` and in **main** I get the following validation error:

salvatore-campagna replied:
I wrote the following **missing** test in **main** not in this branch:

```
---
fetch fields with none stored_fields:
  - do:
      search:
        index: test
        body:
          stored_fields: _none_
          fields: [stored_keyword, keyword, stored_value, value, ignored_keyword, ignored_value, _ignored]

  - match: { hits.total.value: 1 }
```

I see it failing with the following error:
```
[{"error":{"root_cause":[{"type":"action_request_validation_exception","reason":"Validation Failed: 1: [stored_fields] cannot be disabled when using the [fields] option;","stack_trace":"org.elasticsearch.ElasticsearchException$1: Validation Failed: 1: [stored_fields] cannot be disabled when using the [fields] option;
```

Is there anything wrong in my test?

javanna commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java:
```diff
@@ -10,48 +10,102 @@
+                final Set<String> matchingFieldNames = searchExecutionContext.getMatchingFieldNames(storedField);
+                for (final String matchingFieldName : matchingFieldNames) {
+                    final MappedFieldType fieldType = searchExecutionContext.getFieldType(matchingFieldName);
+                    if (SourceFieldMapper.NAME.equals(matchingFieldName) == false
```
this conditional should be before getting the field type: there is no need to retrieve the field type for source and id, and for any other field that is not a metadata field.

javanna commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java:
```diff
@@ -10,48 +10,102 @@
+                    if (SourceFieldMapper.NAME.equals(matchingFieldName) == false
+                        && IdFieldMapper.NAME.equals(matchingFieldName) == false
+                        && searchExecutionContext.isMetadataField(matchingFieldName)
+                        && fieldType.isStored()) {
```
can you add a comment around using `isStored` here to reproduce the previous behaviour we had in `StoredFieldsPhase`. This is purely for bw comp. I think one way to improve this is merging #107086 later.

javanna reviewed (changes requested):
I found two small things that need addressing, but it's real close. Thanks!

salvatore-campagna committed:
fix: remove check on stored_fields for get api

salvatore-campagna committed:
fix: exclude source and id before getting the field type

salvatore-campagna committed:
test: add missing test requesting none stored_fields and a few fields

salvatore-campagna committed:
fix: include a method to know if metadata fields are fetched via wildcard in stored fields phase

salvatore-campagna committed:
Merge branch 'main' into refactor/meta-fields-sub-fetch

salvatore-campagna committed:
fix: skip test in v7 since it was http status 500 instead of 400

salvatore-campagna committed:
Revert "fix: include a method to know if metadata fields are fetched via wildcard in stored fields phase"

This reverts commit 4ace7a02151366fc7336024625e57515ea05741d.

salvatore-campagna committed:
fix: usae HashSet instead of TreeSet

salvatore-campagna committed:
note: clarify why we need to check field is stored

javanna reviewed (approved):
LGTM
